Add quirk to allow downloads on apple.com subdomains

achristensen07 opened the pull request:
#### 2c42b42d082f817f5bd6a3778d42691fd97ef2e9
<pre>
Add quirk to allow downloads on apple.com subdomains
<a href="https://bugs.webkit.org/show_bug.cgi?id=281237">https://bugs.webkit.org/show_bug.cgi?id=281237</a>
<a href="https://rdar.apple.com/136805443">rdar://136805443</a>

Reviewed by Chris Dumez.

As lamentable as it is, there is an internal apple.com subdomain that provides downloads
and has CSP that prevents downloads.  This worked until 283667@main made our behavior more
correct.  The downloads do not work in Chrome or Firefox.  The website owner is working
to resolve the issue, but until it is resolved I have been tasked with making the downloads
continue to work in Safari even though the HTTP header clearly and explicitly prevents downloads.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):

Canonical link: <a href="https://commits.webkit.org/284999@main">https://commits.webkit.org/284999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/64fbc794aecf4cf987d890b2e29cebfdfc4762d6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/50542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/23902 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75236 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22333 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73244 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58342 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22153 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/56705 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/15210 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74194 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/46500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61317 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43163 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/19358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/20674 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/49/builds/58342 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19137 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/76956 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15363 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/18908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15405 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/61358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63932 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12067 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10912 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46342 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/1836 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47413 "Build is in progress. Recent messages:Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/48696 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47155 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1143,7 +1143,10 @@ void DocumentLoader::continueAfterContentPolicy(PolicyAction policy)
 
-        if (!m_frame->effectiveSandboxFlags().contains(SandboxFlag::Downloads)) {
+        // FIXME: Remove this when rdar://137625935 is resolved.
+        bool allowDownloadAnyways = m_frame->isMainFrame() && m_frame->document() && RegistrableDomain(m_frame->protectedDocument()->url()).string() == "apple.com"_s;
```
I think we should use the Quirks class like we usually do.

cdumez commented on Source/WebCore/loader/DocumentLoader.cpp:
```diff
@@ -1143,7 +1143,10 @@ void DocumentLoader::continueAfterContentPolicy(PolicyAction policy)
+        // FIXME: Remove this when rdar://137625935 is resolved.
+        bool allowDownloadAnyways = m_frame->isMainFrame() && m_frame->document() && RegistrableDomain(m_frame->protectedDocument()->url()).string() == "apple.com"_s;
+
+        if (!m_frame->effectiveSandboxFlags().contains(SandboxFlag::Downloads) || allowDownloadAnyways) {
```
Maybe rename `allowDownloadAnyways` to `shouldIgnoreSandboxFlags`?

cdumez reviewed (approved):
LGTM

achristensen07 committed:
Add quirk to allow downloads on apple.com subdomains
https://bugs.webkit.org/show_bug.cgi?id=281237
rdar://136805443

Reviewed by Chris Dumez.

As lamentable as it is, there is an internal apple.com subdomain that provides downloads
and has CSP that prevents downloads.  This worked until 283667@main made our behavior more
correct.  The downloads do not work in Chrome or Firefox.  The website owner is working
to resolve the issue, but until it is resolved I have been tasked with making the downloads
continue to work in Safari even though the HTTP header clearly and explicitly prevents downloads.

* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::continueAfterContentPolicy):

Canonical link: https://commits.webkit.org/284999@main

karlcow commented on Source/WebCore/page/Quirks.cpp:
```diff
@@ -1620,6 +1620,12 @@ bool Quirks::shouldFlipScreenDimensions() const
+bool Quirks::shouldAllowDownloadsInSpiteOfCSP() const
+{
+    // FIXME: Remove this when rdar://137625935 is resolved.
+    return isDomain("apple.com"_s);
```
Could the scope of the quirk be reduced? `apple.com` is very wide. It would be possible to pass the URL and check the path and/or the full domain name.

karlcow replied:
ok after discussing with @achristensen07 let's hope this is **very temporary**.

karlcow reviewed (changes requested):
I'm not comfortable to target all apples subdomains for just one page.
